sklearn/ensemble/tests/test_gradient_boosting.py

```diff
@@ -148,73 +156,88 @@ def test_classification_synthetic():
+    assert_less(error_rate, 0.08)
 
-        gbrt = GradientBoostingClassifier(n_estimators=100, min_samples_split=1,
-                                          max_depth=1, loss=loss,
-                                          learning_rate=1.0, random_state=0)
-        gbrt.fit(X_train, y_train)
-        error_rate = (1.0 - gbrt.score(X_test, y_test))
-        assert error_rate < 0.09, \
-            "GB(loss={}) failed with error {}".format(loss, error_rate)
 
-        gbrt = GradientBoostingClassifier(n_estimators=200, min_samples_split=1,
-                                          max_depth=1,
-                                          learning_rate=1.0, subsample=0.5,
-                                          random_state=0)
-        gbrt.fit(X_train, y_train)
-        error_rate = (1.0 - gbrt.score(X_test, y_test))
-        assert error_rate < 0.08, ("Stochastic GradientBoostingClassifier(loss={}) "
-                                   "failed with error {}".format(loss, error_rate))
+def test_classification_synthetic():
+    for presort, loss in product((True, False), ('deviance', 'exponential')):
+        yield check_classification_synthetic, presort, loss
 
 
-def test_boston():
+def check_boston(presort, loss, subsample):
     # Check consistency on dataset boston house prices with least squares
     # and least absolute deviation.
-    for loss in ("ls", "lad", "huber"):
-        for subsample in (1.0, 0.5):
-            last_y_pred = None
-            for i, sample_weight in enumerate(
-                    (None, np.ones(len(boston.target)),
-                     2 * np.ones(len(boston.target)))):
-                clf = GradientBoostingRegressor(n_estimators=100, loss=loss,
-                                                max_depth=4, subsample=subsample,
-                                                min_samples_split=1,
-                                                random_state=1)
-
-                assert_raises(ValueError, clf.predict, boston.data)
-                clf.fit(boston.data, boston.target,
-                        sample_weight=sample_weight)
-                leaves = clf.apply(boston.data)
-                assert_equal(leaves.shape, (506, 100))
-                
-                y_pred = clf.predict(boston.data)
-                mse = mean_squared_error(boston.target, y_pred)
-                assert mse < 6.0, "Failed with loss %s and " \
-                    "mse = %.4f" % (loss, mse)
-
-                if last_y_pred is not None:
-                    np.testing.assert_array_almost_equal(
-                        last_y_pred, y_pred,
-                        err_msg='pred_%d doesnt match last pred_%d for loss %r and subsample %r. '
-                        % (i, i - 1, loss, subsample))
-
-                last_y_pred = y_pred
+    last_y_pred = None
+    for i, sample_weight in enumerate(
+            (None, np.ones(len(boston.target)),
+             2 * np.ones(len(boston.target)))):
```

Comment:
nitpick: Strange indentation.

I would do 
```
or i, sample_weight in enumerate((..., 
                                  ...,
                                  ...)):
```

or define the list of sample weight before the loop.